File: 2015/cpp/day03.cpp
#include <cassert>
#include <string>
#include <set>
#include <fstream>
#include <cstdio>

using namespace std;

auto Problem1(const std::string &input) -> int {
    set<pair<int, int>> s;
    s.insert({0, 0});

    int row = 0;
    int col = 0;
    for (char c : input) {
        switch (c) {
        case '^':
            row -= 1;
            break;
        case '>':
            col += 1;
            break;
        case 'v':
            row += 1;
            break;
        case '<':
            col -= 1;
            break;
        default:
            assert(!"never reach here");
            break;
        }
        s.insert({row, col});
    }

    return s.size();
}

auto Problem2(const std::string &input) -> int {
    set<pair<int, int>> s;
    s.insert({0, 0});

    int rows[2] = {0, 0};
    int cols[2] = {0, 0};
    int index = 0;
    for (char c : input) {
        switch (c) {
        case '^':
            rows[index] -= 1;
            break;
        case '>':
            cols[index] += 1;
            break;
        case 'v':
            rows[index] += 1;
            break;
        case '<':
            cols[index] -= 1;
            break;
        default:
            assert(!"never reach here");
            break;
        }

        s.insert({rows[index], cols[index]});
        index = index == 0 ? 1 : 0;
    }

    return s.size();
}

void Test() {
    assert(Problem1(">") == 2);
    assert(Problem1("^>v<") == 4);
    assert(Problem1("^v^v^v^v^v") == 2);

    assert(Problem2("^v") == 3);
    assert(Problem2("^>v<") == 3);
    assert(Problem2("^v^v^v^v^v") == 11);
}

int main() {
    Test();

    ifstream f("../input/day03.txt");
    string input;
    f >> input;

    int ret1 = Problem1(input);
    assert(ret1 == 2572);

    int ret2 = Problem2(input);
    assert(ret2 == 2631);

    printf("ret1 = %d\n", ret1);
    printf("ret2 = %d\n", ret2);
    return 0;
}

File: 2015/python/day03.py
#!/usr/bin/env python3


def part01(input: str) -> int:
    s: set[tuple[int, int]] = set()
    s.add((0, 0))

    x = 0
    y = 0
    for c in input:
        if c == "^":
            y += 1
        elif c == ">":
            x += 1
        elif c == "v":
            y -= 1
        elif c == "<":
            x -= 1

        s.add((x, y))

    return len(s)


def part02(input: str) -> int:
    s: set[tuple[int, int]] = set()
    s.add((0, 0))

    is_human = True
    x1 = 0
    y1 = 0
    x2 = 0
    y2 = 0
    for c in input:
        if is_human:
            if c == "^":
                y1 += 1
            elif c == ">":
                x1 += 1
            elif c == "v":
                y1 -= 1
            elif c == "<":
                x1 -= 1
            is_human = False
            s.add((x1, y1))
        else:
            if c == "^":
                y2 += 1
            elif c == ">":
                x2 += 1
            elif c == "v":
                y2 -= 1
            elif c == "<":
                x2 -= 1
            is_human = True
            s.add((x2, y2))

    return len(s)


with open("../input/day03.txt") as f:
    input = f.readline()

    ret1 = part01(input)
    assert ret1 == 2572

    ret2 = part02(input)
    assert ret2 == 2631

    print(f"ret1={ret1}")
    print(f"ret2={ret2}")
